text/0005-sapper-as-middleware.md

```diff
@@ -0,0 +1,165 @@
+export (at least) two functions:
+
+- A `middleware(req, res)` function that will serve a dynamic request.
+- A `assets(req, res)` function that will serve a static asset request.
+
+The supporting files *must* also include a manifest of files that can be uploaded to a CDN.
+
+This javascript *must not* start an Express server.
+
+## Motivation
+
+Working with Sapper in production for a large scale project has identified a number of
+challenges that arise from the unintended complexity of stale assumptions.
+
+By reducing these assumptions to the *absolute minumum*, we aim to:
+
+- Make it easier to deploy Sapper in efficient, scalable configurations
+- Make it easy to use Sapper applications in more scenarios than just a single standalone site
+- Make it easier to extend Sapper to suit specific project needs (without requiring modifications of Sapper itself)
+
+As of 0.28, the Sapper build process emits a completely standalone directory in `__sapper__/build`
+that starts an Express server listening on `$PORT` to serve the Sapper site.
+
+To do this Sapper makes many nuanced assumptions about the dependencies and runtime environment of
+applications it is building. Many of these assumptions are unnecessary sources of complexity during
+both developent and deployment.
+
+## Detailed design
+
+The general behavior of client-side bundling for Sapper should remain unchanged. Components are compiled for
+each route as SSR code and as DOM code. The behavior of the DOM (client-side) entry point is also unchanged.
+We expect the client-side bundle to be just as completely self-contained as it is today.
+
+This is in contrast with the server-side bundling operation. Instead of emitting a completely free-standing
+NodeJS script that starts an Express server, it would instead emit a partially bundled .js file that the user
+should include in a script of their own. This server-side .js file would look something like:
+
+```javascript
+
+export function middleware(req, res) {
+  // Sapper routing logic here to handle dynamic requests ...
+}
+
+export function assets(req, res) {
+  // sirv-based logic here to serve requests for static assets ...
+}
+
+```
+
+There must also be a manifest that's emitted to make deployment as straightforward as possible. A simple
+manifest might look something like:
+
+```json
+{
+  "assets": {
+    "directory": "assets",
+    "entries": {
+      "client/index.5bdaeb21.js": {
+        "headers": [
+          {
+            "key": "Cache-Control",
+            "value": "max-age=31536000, immutable"
+          },
+          {
+            "key": "Content-Type",
+            "value": "application/javascript"
+          }
+        ]
+      },
+      "index.css": {
+        "headers": [
+          {
+            "key": "Content-Type",
+            "value": "text/css"
+          }
+        ]
+      }
+    }
+  },
+  "functions": {
+    "index.js": {}
```

Comment:
what would the right-hand side of this contain?

Reply:
Probably details about the requirements of this function.

Things like environment variables, minimum memory, required node versions.

Reply:
Is that something I want Sapper to control though? How would it know what to put there? It seems like something I need to specify as a developer

Reply:
Maybe just the search path would be enough to start with? I agree with @benmccann that platform specific requirements like CPU, memory, etc should come from elsewhere.